src/utils/daos_hdlr.c

```diff
@@ -1535,6 +1536,741 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+daos_file_stat(daos_file_t *daos_file, const char *path, struct stat *buf)
+{
+	int rc = 0;
+	if (daos_file->type == POSIX) {
+		rc = stat(path, buf);
+	} else if (daos_file->type == DAOS) {
+		rc = daos_stat(daos_file, path, buf);
+	} else {
+		fprintf(stderr, "File type not known, file=%s, type=%d",
+			path, daos_file->type);
+	}
+	return rc;
+}
+
+static ssize_t
+daos_read(daos_file_t *daos_file,
+	const char* file,
+	void* buf,
+	size_t size)
+{
+	d_iov_t iov;
+	d_iov_set(&iov, buf, size);
+
+	/* define scatter-gather list for dfs_write */
+	d_sg_list_t sgl;
+	sgl.sg_nr = 1;
+	sgl.sg_iovs = &iov;
+	sgl.sg_nr_out = 1;
+
+	/* execute read operation */
+	daos_size_t got_size;
+	int rc = dfs_read(daos_file->dfs, daos_file->obj, &sgl,
+			daos_file->offset, &got_size, NULL); 
+	if (rc) {
+		fprintf(stderr, "dfs_read %s failed (%d %s)",
+			file, rc, strerror(rc));
+		errno = rc;
+		return -1;
+	}
+	/* update file pointer with number of bytes read */
+	daos_file->offset += (daos_off_t)got_size;
+
+	return (ssize_t)got_size;
+}
+
+static ssize_t
+daos_file_read(struct daos_file_t *daos_file, char* file,
+		void* buf, size_t size)
+{
+	ssize_t got_size = 0;
+	if (daos_file->type == POSIX) {
+		got_size = read(daos_file->fd, buf, size);
+	} else if (daos_file->type == DAOS) {
+		got_size = daos_read(daos_file, file, buf, size);
+	} else {
+		fprintf(stderr, "File type not known: %s type=%d", file, daos_file->type);
+	}
+	return got_size;
+}
+
+static int
+daos_closedir(DIR* _dirp)
+{
+	struct dfs_daos_t *dirp = (struct dfs_daos_t *)_dirp;
+	int rc = dfs_release(dirp->dir);
+	if (rc) {
+		fprintf(stderr, "dfs_release failed (%d %s)",
+			rc, strerror(rc));
+		errno = rc;
+		rc = -1;
+	}
+	free(dirp);
+	return rc;
+}
+
+static int
+daos_file_closedir(daos_file_t *daos_file, DIR* dirp)
+{
+	int rc = 0;
+	if (daos_file->type == POSIX) {
+		rc = closedir(dirp);
+	} else if (daos_file->type == DAOS) {
+		rc = daos_closedir(dirp);
+	} else {
+		fprintf(stderr, "File type not known, type=%d", daos_file->type);
+	}
+	return rc;
+}
+
+static int
+daos_close(daos_file_t *daos_file, const char* file)
+{
+	int rc = dfs_release(daos_file->obj);
+	if (rc) {
+		fprintf(stderr, "dfs_close %s failed (%d %s)",
+			file, rc, strerror(rc));
+		errno = rc;
+		rc = -1;
+	}
+	return rc;
+}
+
+static int
+daos_file_close(daos_file_t *daos_file, const char *file)
+{
+	int rc = 0;
+	if (daos_file->type == POSIX) {
+		rc = close(daos_file->fd);
+		if (rc == 0)
+			daos_file->fd = -1;
+	} else if (daos_file->type == DAOS) {
+		rc = daos_close(daos_file, file);
+		if (rc == 0)
+			daos_file->obj = NULL;
+	} else {
+		fprintf(stderr, "File type not known, file=%s, type=%d",
+			file, daos_file->type);
+	}
+	return rc;
+}
+
+static void
+fs_copy(daos_file_t *daos_src_file,
+		daos_file_t *daos_dst_file,
+		const char* dir_name,
+		const char* fs_dst_prefix)
+{
+	DIR *d;
+	int rc;
+
+	/* begin by opening source directory */
+	d = daos_file_opendir(daos_src_file, dir_name);
```

Comment:
To be clear, we are only allowing a source directory, not a source file?

Reply:
@daltonbohning yes, in the case that it moving data from a DAOS container to a source file I am trying to think about if that would make sense. The DAOS source would need to be a single file, which is a case we should probably handle.

Reply:
I think we probably should, or at least error. But maybe we can do that in a future iteration?

Reply:
@daltonbohning yeah, for now we can error. This PR is already fairly large so I can add support for that in the next iteration.

Reply:
I agree. And once we have a test plan and start implementing tests we can hammer out some of the finer points like this.

Reply:
@daltonbohning ok, I've added an error check here if the source is not a directory, we can also work out better in which cases we want to return from this function

Reply:
> which cases we want to return from this function

Do you mean to do this on this PR, or future? I'm okay with a future PR, since we have some other finer points to handle as well.

Reply:
@daltonbohning in another PR, thanks for all of your feedback

Reply:
You're welcome!